tests/test_install.sh

```diff
@@ -0,0 +1,157 @@
+}
+
+testArtifactInstallConfigExists() {
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    assertEquals '{"new":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
+testNeedsArtifactRebootAutomatic() {
+    echo 'exit 20' >> "${TEST_APPLY_CMD}"
+
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    output="$(./$mender_configure NeedsArtifactReboot "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "Automatic" "${output}"
+
+    assertEquals '{"new":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
+testNeedsArtifactRebootNo() {
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    output="$(./$mender_configure NeedsArtifactReboot "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "No" "${output}"
+
+    assertEquals '{"new":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
+testArtifactCommit() {
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    output="$(./$mender_configure NeedsArtifactReboot "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "No" "${output}"
+
+    output="$(./$mender_configure ArtifactCommit "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    assertEquals '{"new":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
+testArtifactRollback() {
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    output="$(./$mender_configure NeedsArtifactReboot "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "No" "${output}"
+
+    output="$(./$mender_configure ArtifactRollback "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    assertEquals '{"old":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}
+apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
+testArtifactRollbackAndReboot() {
+    echo 'if [ "$TEST_ROLLBACK" = 1 ]; then exit 20; fi' >> "${TEST_APPLY_CMD}"
+
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    output="$(./$mender_configure NeedsArtifactReboot "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "No" "${output}"
+
+    export TEST_ROLLBACK=1
+    output="$(./$mender_configure ArtifactRollback "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+    unset TEST_ROLLBACK
+
```

Comment:
How is different from the previous test case? To actually verify the Rollback+Reboot case we need to call here `NeedsArtifactReboot` again and verify it now says `Automatic`.

Reply:
You're right, I should test `NeedsArtifactReboot` again here. Good catch.

The test case is a little bit contrived, as I don't think the client actually calls `NeedsArtifactReboot` again after the first call, but I wanted to test that at least the return code 20 was handled correctly, also in `ArtifactRollback`.